src/src/com/microsoft/aad/adal/HttpWebRequest.java

```diff
@@ -107,107 +89,103 @@ private void setupConnection() {
         }
         HttpURLConnection.setFollowRedirects(true);
-        mConnection = openConnection();
+        final HttpURLConnection connection = (HttpURLConnection)mUrl.openConnection();
+        connection.setConnectTimeout(CONNECT_TIME_OUT);
         // To prevent EOF exception.
         if (Build.VERSION.SDK_INT > 13) {
-            mConnection.setRequestProperty("Connection", "close");
+            connection.setRequestProperty("Connection", "close");
         }
+
+        // Apply the request headers
+        final Iterator<String> headerKeys = mRequestHeaders.keySet().iterator();
+
+        while (headerKeys.hasNext()) {
+            String header = headerKeys.next();
+            Logger.v(TAG, "Setting header: " + header);
+            connection.setRequestProperty(header, mRequestHeaders.get(header));
+        }
+
+        // Avoid reuse of existing sockets to avoid random EOF errors
+        System.setProperty("http.keepAlive", "false");
+        connection.setReadTimeout(READ_TIME_OUT);
+        connection.setInstanceFollowRedirects(mInstanceRedirectsFollow);
+        connection.setUseCaches(mUseCaches);
+        connection.setRequestMethod(mRequestMethod);
+        connection.setDoInput(true); // it will at least read status
+                                     // code. Default is true.
+        setRequestBody(connection);
+
+        return connection;
     }
 
     /**
      * send the request.
-     * 
-     * @param contentType
      */
-    public HttpWebResponse send() {
-
-        Logger.v(TAG, "HttpWebRequest send thread:" + android.os.Process.myTid());
-        setupConnection();
-        HttpWebResponse response = new HttpWebResponse();
+    public HttpWebResponse send() throws IOException {
+        Logger.v(TAG, "HttpWebRequest send thread:" + Process.myTid());
+        final HttpURLConnection connection = setupConnection();
+        if (connection == null) {
+            return null;
+        }
 
-        if (mConnection != null) {
+        final HttpWebResponse response;
+        try {
+            InputStream responseStream = null;
             try {
-                // Apply the request headers
-                final Iterator<String> headerKeys = mRequestHeaders.keySet().iterator();
-
-                while (headerKeys.hasNext()) {
-                    String header = headerKeys.next();
-                    Logger.v(TAG, "Setting header: " + header);
-                    mConnection.setRequestProperty(header, mRequestHeaders.get(header));
+                responseStream = connection.getInputStream();
+            } catch (IOException ex) {
+                Logger.e(TAG, "IOException:" + ex.getMessage(), "", ADALError.SERVER_ERROR);
+                // If it does not get the error stream, it will return
+                // exception in the httpresponse
+                responseStream = connection.getErrorStream();
+                if (responseStream == null) {
+                    throw ex;
```

Comment:
Nope, we should check error stream which contains data why refresh token failed...